src/upload/form/project_assets.rs

```diff
@@ -81,64 +87,166 @@ impl ServiceWorkerAssets {
     }
 }
 
+#[derive(Debug, PartialEq, PartialOrd, Eq, Ord)]
 pub struct Module {
     pub name: String,
     pub path: PathBuf,
     pub module_type: ModuleType,
 }
 
-impl Module {
-    pub fn new(name: String, path: PathBuf) -> Result<Module, failure::Error> {
-        let extension = path
-            .extension()
-            .ok_or_else(|| {
-                failure::err_msg(format!(
-                    "File {} lacks an extension. An extension is required to determine module type",
-                    path.display()
-                ))
-            })?
-            .to_string_lossy();
-
-        let module_type = match extension.as_ref() {
-            "mjs" => ModuleType::ES6,
-            "js" => ModuleType::CommonJS,
-            "wasm" => ModuleType::Wasm,
-            "txt" => ModuleType::Text,
-            "bin" => ModuleType::Data,
-            unknown => failure::bail!(format!(
-                "unknown extension {}, cannot determine module type",
-                unknown
-            )),
-        };
-
-        Ok(Module {
-            name,
-            path,
-            module_type,
-        })
-    }
-}
-
+#[derive(Clone, Copy, Debug, PartialEq, PartialOrd, Eq, Ord)]
 pub enum ModuleType {
-    ES6,
+    ESModule,
     CommonJS,
-    Wasm,
+    CompiledWasm,
     Text,
     Data,
 }
 
 impl ModuleType {
     pub fn content_type(&self) -> &str {
         match &self {
-            Self::ES6 => "application/javascript+module",
+            Self::ESModule => "application/javascript+module",
             Self::CommonJS => "application/javascript",
-            Self::Wasm => "application/wasm",
+            Self::CompiledWasm => "application/wasm",
             Self::Text => "text/plain",
             Self::Data => "application/octet-stream",
         }
     }
 }
 
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct ModuleGlobs {
```

Comment:
Should there be a catch-all `exclude` glob list?

Reply:
unknown files are ignored now by default, and you can exclude files that would match an individual rule with a negative glob rule I believe (I'll confirm this in a test.